### PR TITLE
Skeleton for binary parser

### PR DESCRIPTION
**Pull request by hjorthjort**
This PR sets up the basics of the binary-format to KWasm parser.

* Add py-wasm as a dependency.
* Use py-wasm to parse binary format into an abstract format in Python.
* Translate abstract format to KWasm using pyk (only for types and functions without bodies, at the moment).
* Test the translation by running the modules in pyk.

### CODE REVIEW

**Comment by hjorthjort** [comment]
The `revert later` comment is not yet relevant, since we don't have what we need to support the delegation contract yet.

File: Makefile [hunk 1]
```diff
@@ -42,6 +49,9 @@ K_JAR := $(K_SUBMODULE)/k-distribution/target/release/k/lib/java/kernel-1.0-SNAP
 
 deps: $(K_JAR) $(TANGLER)
 
+$(PYWASM_DEPS_PATH):
+	cd $(PYWASM_PATH) && virtialenv -p python3 venv && . venv/bin/activate
```
Comment (hjorthjort) [inline Makefile:53]:
```suggestion
	cd $(PYWASM_PATH) && virtualenv -p python3 venv && . venv/bin/activate
```

File: Dockerfile [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+RUN pip3 install virtualenv
+
+RUN    git clone 'https://github.com/WebAssembly/wabt' --branch 1.0.13 --recurse-submodules wabt \
+    && cd wabt                                                                                   \
+    && mkdir build                                                                               \
+    && cd build                                                                                  \
+    && cmake ..                                                                                  \
+    && cmake --build .
+
+ENV PATH=/home/user/wabt/build:$PATH
```
Comment (hjorthjort) [inline Dockerfile:29]:
```suggestion
ENV PATH=/home/user/wabt/build:/home/user/.local/bin:$PATH
```

File: Makefile [hunk 3]
```diff
@@ -24,9 +24,16 @@ endif
 PATH := $(K_BIN):$(PATH)
 export PATH
 
+PYK_PATH         := $(K_LIB)
+PYWASM_PATH      := ./deps/py-wasm
+PYWASM_DEPS_PATH := $(PYWASM_PATH)/venv/lib/python3.6/site-packages
+
+PYTHONPATH := $(PYK_PATH):$(PYWASM_PATH):$(PYWASM_DEPTHS_PATH):$(PYTHONPATH)
```
Comment (hjorthjort) [inline Makefile:31]:
```suggestion
PYTHONPATH := $(PYK_PATH):$(PYWASM_PATH):$(PYWASM_DEPS_PATH):$(PYTHONPATH)
```

File: Makefile [hunk 4]
```diff
@@ -24,9 +24,16 @@ endif
 PATH := $(K_BIN):$(PATH)
 export PATH
 
+PYK_PATH         := $(K_LIB)
+PYWASM_PATH      := ./deps/py-wasm
+PYWASM_DEPS_PATH := $(PYWASM_PATH)/venv/lib/python3.6/site-packages
```
Comment (hjorthjort) [inline Makefile:29]:
```suggestion
PYWASM_DEPS_PATH := $(dir $(shell find $(PYWASM_PATH) -name os.py))
```

File: Makefile [hunk 5]
```diff
@@ -42,6 +49,9 @@ K_JAR := $(K_SUBMODULE)/k-distribution/target/release/k/lib/java/kernel-1.0-SNAP
 
 deps: $(K_JAR) $(TANGLER)
 
+$(PYWASM_DEPS_PATH):
```
Comment (hjorthjort) [inline Makefile:52]:
```suggestion
$(PYWAS_PATH)/lib:
```

File: Makefile [hunk 6]
```diff
@@ -189,6 +199,30 @@ proof_tests:=$(wildcard tests/proofs/*-spec.k)
...
 test-prove: $(proof_tests:=.prove)
 
+### Binary Parser Test
+
+BINARY:=python3 binary-parser/test.py
+
+tests/binary/%.wasm: tests/binary/%.wat
+	wat2wasm $< --output=$@
+
+tests/%.wasm.bparse: tests/%.wasm  $(PYWASM_DEPS_PATH)
```
Comment (hjorthjort) [inline Makefile:209]:
```suggestion
tests/%.wasm.bparse: tests/%.wasm  $(PYWASM_PATH)/lib
```

File: Makefile [hunk 7]
```diff
@@ -189,6 +199,30 @@ proof_tests:=$(wildcard tests/proofs/*-spec.k)
...
 test-prove: $(proof_tests:=.prove)
 
+### Binary Parser Test
+
+BINARY:=python3 binary-parser/test.py
+
+tests/binary/%.wasm: tests/binary/%.wat
+	wat2wasm $< --output=$@
+
+tests/%.wasm.bparse: tests/%.wasm  $(PYWASM_PATH)/lib
```
Comment (hjorthjort) [inline Makefile:209]:
```suggestion
tests/%.wasm.bparse: tests/%.wasm  $(PYWASM_PATH)/venv
```

File: Makefile [hunk 8]
```diff
@@ -42,6 +49,9 @@ K_JAR := $(K_SUBMODULE)/k-distribution/target/release/k/lib/java/kernel-1.0-SNAP
 
 deps: $(K_JAR) $(TANGLER)
 
+$(PYWAS_PATH)/lib:
```
Comment (hjorthjort) [inline Makefile:52]:
```suggestion
$(PYWAS_PATH)/venv:
```

File: Makefile [hunk 9]
```diff
@@ -42,6 +49,9 @@ K_JAR := $(K_SUBMODULE)/k-distribution/target/release/k/lib/java/kernel-1.0-SNAP
 
 deps: $(K_JAR) $(TANGLER)
 
+$(PYWAS_PATH)/venv:
```
Comment (hjorthjort) [inline Makefile:52]:
```suggestion
$(PYWASM_PATH)/venv:
```

File: Dockerfile [hunk 10]
```diff
@@ -1,14 +1,29 @@
...
+
 
 ARG USER_ID=1000
 ARG GROUP_ID=1000
 RUN groupadd -g $GROUP_ID user && useradd -m -u $USER_ID -s /bin/sh -g user user
 
 USER user:user
 WORKDIR /home/user
+
+RUN pip3 install virtualenv
```
Comment (hjorthjort) [inline Dockerfile:20]:
```suggestion
RUN pip3 install numpy cytoolz
```

File: Makefile [hunk 11]
```diff
@@ -42,6 +49,9 @@ K_JAR := $(K_SUBMODULE)/k-distribution/target/release/k/lib/java/kernel-1.0-SNAP
 
 deps: $(K_JAR) $(TANGLER)
 
+$(PYWASM_PATH)/venv:
+	cd $(PYWASM_PATH) && virtualenv -p python3 venv && . venv/bin/activate
```
Comment (hjorthjort) [inline Makefile:53]:
```suggestion
	cd $(PYWASM_PATH) && virtualenv -p python3 venv
```

File: Makefile [hunk 12]
```diff
@@ -189,6 +199,30 @@ proof_tests:=$(wildcard tests/proofs/*-spec.k)
 
 test-prove: $(proof_tests:=.prove)
 
+### Binary Parser Test
+
+BINARY:=python3 binary-parser/test.py
```
Comment (hjorthjort) [inline Makefile:204]:
```suggestion
BINARY:=cd $(PYWASM_PATH) && . venv/bin/activate && python3 binary-parser/test.py
```

**Comment by hjorthjort** [comment]
@ehildenb I'm trying to get the virtual environment setup to work, so that all the right modules are available when running the binary parser. Any ideas?

File: binary-parser/kwasm_ast.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
+from pyk.kast import KSequence, KConstant, KApply, KToken
```
Comment (ehildenb) [inline binary-parser/kwasm_ast.py:1]:
```suggestion
#!/usr/bin/env python3

from pyk.kast import KSequence, KConstant, KApply, KToken
```

File: binary-parser/test.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import wasm2kast
```
Comment (ehildenb) [inline binary-parser/test.py:1]:
```suggestion
#!/usr/bin/env python3

import wasm2kast
```

File: binary-parser/wasm2kast.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import sys
```
Comment (ehildenb) [inline binary-parser/wasm2kast.py:1]:
```suggestion
#!/usr/bin/env python3

import sys
```

**Review by ehildenb — APPROVED** [review]

**Pull request merged**